src/main/java/org/openstreetmap/atlas/checks/validation/linear/lines/WaterWayCheck.java

```diff
@@ -0,0 +1,453 @@
+                l -> l.asPolyLine().contains(last) && !last.equals(l.asPolyLine().last()));
+    }
+
+    /**
+     * Check if a location is to the right of a line (e.g., kerbs, cliffs, and oceans)
+     *
+     * @param line
+     *            The line to compare the location to
+     * @param location
+     *            The location
+     * @return {@code true} if the location is to the right of the line. If the location is on the
+     *         line, or to the left of the line, we return {@code false}.
+     */
+    public boolean isRightOf(final PolyLine line, final Location location)
+    {
+        final PolyLine tLine = new PolyLine(location);
+        final Segment closest = line.segments().stream()
+                .min(Comparator.comparingDouble(s -> s.shortestDistanceTo(tLine).asMeters()))
+                .orElse(null);
+        if (closest != null)
+        {
+            final PolyLine testLine = new PolyLine(closest.first(), closest.last(), location);
+            final Angle difference = testLine.headingDifference().orElse(null);
+            if (difference != null && difference.asDegrees() > 0)
+            {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Check if a location is inside an atlas bounds
+     *
+     * @param atlas
+     *            The atlas
+     * @param location
+     *            The location to check
+     * @return {@code true} if the atlas contains the location inside its bounds
+     */
+    public boolean isValidEndToCheck(final Atlas atlas, final Location location)
+    {
+        return atlas.bounds().fullyGeometricallyEncloses(location);
+    }
+
+    @Override
+    public boolean validCheckForObject(final AtlasObject object)
+    {
+        return !isFlagged(object.getOsmIdentifier()) && object instanceof LineItem
+                && this.waterwayTagFilter.test(object);
+    }
+
+    @Override
+    protected Optional<CheckFlag> flag(final AtlasObject object)
+    {
+        final LineItem line = (LineItem) object;
+        final Location last = line.asPolyLine().last();
+        final Location first = line.asPolyLine().first();
+        CheckFlag flag = null;
+        final double incline = this.elevationUtils.getIncline(first, last);
+        final boolean uphill = !Double.isNaN(incline) && incline > 0
+                && last.distanceTo(first).isGreaterThan(this.minDistanceStartEndElevationUphill);
+        if (line.isClosed())
+        {
+            flag = createFlag(object,
+                    this.getLocalizedInstruction(FALLBACK_INSTRUCTIONS.indexOf(CIRCULAR_WATERWAY),
+                            object.getOsmIdentifier()),
+                    Collections.singletonList(line.asPolyLine().first()));
+        }
+        else if (uphill && this.minResolutionDistance
+                .isGreaterThanOrEqualTo(this.elevationUtils.getResolution(first)))
+        {
+            flag = createUphillFlag(object, first);
```

Comment:
We probably want to add that this check flags uphill waterways as well (docs/javadoc)
I noticed it in a couple places, but wasn't sure if it was consistently mentioned

Reply:
I'll do that (I wrote the docs, then wrote the test, then checked NZL, where there are _many_ waterways that are going the wrong direction, and _then_ wrote the elevation check code).